### PR TITLE
Propose relationships design

### PR DESCRIPTION
**Pull request by ajmaddaford**
## What is the context of this PR?
This pull request proposes a design for relationships using the new [url structure](0005-simplify-urls-and-extend-to-support-repeating-effectively.md) and [lists](0006-make-named-lists-a-first-class-construct.md).

## How to review
- Is there enough detail, or too much?
- Is there a better approach?
- Is anything missing?

### CODE REVIEW

File: doc/architecture/decisions/0007-relationships.md [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+  "identifiers": ["first-name", "last-name"],
+}
+```
+
+The second person name is more difficult, as we will need to resolve the answers for a list item outside of the current repeat. We could introduce a `list_item_id` property to the placeholder source definition. This could be a fixed value that can be used to resolve a list item id. In this example runner would need to resolve `related_list_item_id` to the url parameter:
+
+```
+{
+  "source": "answers",
+  "list_item_id": "related_list_item_id"
```
Comment (jonnyshaw89) [inline doc/architecture/decisions/0007-relationships.md]: I think we need a more explicit way to say that this is a value from a url parameter?

**Review by jonnyshaw89 — CHANGES_REQUESTED** [review]
Mostly happy with this proposal

File: doc/architecture/decisions/0007-relationships.md (hunk 1 above)
Comment (ajmaddaford) [reply doc/architecture/decisions/0007-relationships.md]: It's not as far as the schema is concerned. Getting the value from a url parameter is how we plan to do it in the runner implementation.

Comment (jonnyshaw89) [reply doc/architecture/decisions/0007-relationships.md]: I think if we make the value less like a string value and maybe prefix with to make it look like a pointer?

Comment (ajmaddaford) [reply doc/architecture/decisions/0007-relationships.md]: Thinking about this, `list_item_id` will never be an actual list item id, as the list items will only exist at runtime. We could either use a prefix (e.g. @):

```
{
  "source": "answers",
  "list_item_id": "@related_list_item_id"
  "identifiers": ["first-name", "last-name"],
}
```

Or change the name of the property, to something like `list_item_selector`:
```
{
  "source": "answers",
  "list_item_selector": "related_list_item_id"
  "identifiers": ["first-name", "last-name"],
}
```

Comment (jonnyshaw89) [reply doc/architecture/decisions/0007-relationships.md]: i think the latter is more appropriate

Comment (bitdivision) [reply doc/architecture/decisions/0007-relationships.md]: I agree, selector makes sense immediately, whereas you would need to think about the `@` if you didn't know anything about it.

**Review by MebinAbraham — APPROVED** [review]

**Review by bitdivision — APPROVED** [review]

**Review by jonnyshaw89 — APPROVED** [review]

**Pull request merged**